File: code/05-tail-call/hello-tail.c
// hello-tail.c — eBPF 尾调用实验
//
// 本程序演示:
//   1. 用 BPF_MAP_TYPE_PROG_ARRAY 实现尾调用
//   2. 程序链：entry -> handler1 -> handler2
//   3. fallback 降级机制
//
// 编译: clang -O2 -target bpf -c hello-tail.c -o hello-tail.o

#include <uapi/linux/ptrace.h>
#include <linux/bpf.h>
#include <linux/version.h>
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>

// ============================================================
// 1. 定义 prog_array map（尾调用的核心）
//    key: u32 索引
//    value: u32 fd（程序文件描述符）
//    max_entries: 链中最多几个程序
// ============================================================
struct {
    __uint(type, BPF_MAP_TYPE_PROG_ARRAY);
    __uint(max_entries, 8);
    __type(key, __u32);
    __type(value, __u32);
} jmp_table SEC(".maps");

// ============================================================
// 2. 共享数据 map（tail call 之间传递数据）
// ============================================================
struct {
    __uint(type, BPF_MAP_TYPE_ARRAY);
    __uint(max_entries, 1);
    __type(key, __u32);
    __type(value, __u64);
} shared_data SEC(".maps");

// ============================================================
// 3. 许可证
// ============================================================
char LICENSE[] SEC("license") = "GPL";

// ============================================================
// 程序 1: 入口 (entry)
// 挂载点: kprobe/sys_execve
// 功能: 收集基本信息，分派到处理程序
// ============================================================
SEC("kprobe/sys_execve")
int hello_entry(struct pt_regs *ctx) {
    u64 pid = bpf_get_current_pid_tgid();
    u32 key = 0;

    bpf_trace_printk("[ENTRY] pid=%u, entering tail call chain\\n",
                     pid >> 32);

    // 保存 pid 到共享 map，给后续程序用
    bpf_map_update_elem(&shared_data, &key, &pid, BPF_ANY);

    // 尾调用到索引 1（hello_handler1）
    // 关键: bpf_tail_call 成功则不返回！
    bpf_tail_call(ctx, &jmp_table, 1);

    // 如果执行到这里，说明尾调用失败（handler1 未加载）
    bpf_trace_printk("[FALLBACK] tail call to handler1 failed\\n");

    return 0;
}

// ============================================================
// 程序 2: 处理器 1 (handler1)
// 功能: 进程信息处理，继续尾调用
// ============================================================
SEC("kprobe/sys_execve")
int hello_handler1(struct pt_regs *ctx) {
    u64 pid = bpf_get_current_pid_tgid();
    char comm[16] = {};
    u32 key = 0;

    bpf_get_current_comm(&comm, sizeof(comm));

    bpf_trace_printk("[HANDLER1] pid=%u comm=%s processing...\\n",
                     pid >> 32, comm);

    // 继续尾调用到索引 2（hello_handler2）
    bpf_tail_call(ctx, &jmp_table, 2);

    // 如果 handler2 未加载，走降级
    bpf_trace_printk("[HANDLER1] no handler2, completing\\n");
    return 0;
}

// ============================================================
// 程序 3: 处理器 2 (handler2)
// 功能: 最终处理，链结束
// ============================================================
SEC("kprobe/sys_execve")
int hello_handler2(struct pt_regs *ctx) {
    u64 pid = bpf_get_current_pid_tgid();
    char comm[16] = {};
    u32 key = 0;

    bpf_get_current_comm(&comm, sizeof(comm));
    u64 *shared_pid = bpf_map_lookup_elem(&shared_data, &key);

    bpf_trace_printk("[HANDLER2] pid=%u comm=%s end of chain\\n",
                     pid >> 32, comm);

    if (shared_pid) {
        bpf_trace_printk("[HANDLER2] entry saved pid=%u\\n",
                         (u32)(*shared_pid >> 32));
    }

    return 0;
}

// ============================================================
// 程序 4: 降级处理器 (fallback)
// 当主链出问题时使用
// ============================================================
SEC("kprobe/sys_execve")
int hello_fallback(struct pt_regs *ctx) {
    u64 pid = bpf_get_current_pid_tgid();

    bpf_trace_printk("[FALLBACK] pid=%u basic monitoring only\\n",
                     pid >> 32);
    return 0;
}

File: code/05-tail-call/hello-tail.py
#!/usr/bin/python3
# hello-tail.py — eBPF 尾调用实验加载器
#
# 演示:
#   1. 加载多个 eBPF 程序到内核
#   2. 用 prog_array 建立程序链
#   3. 动态切换尾调用目标
#   4. bpftool 观察尾调用状态
#
# 用法:
#   sudo python3 hello-tail.py              # 完整 3 级链
#   sudo python3 hello-tail.py --chain=2    # 只链 2 级（观察 fallback）
#   sudo python3 hello-tail.py --no-fallback # 不加载 fallback（观察错误行为）

import sys
import subprocess
import os
from bcc import BPF

# 确保以 root 运行
if os.geteuid() != 0:
    print("❌ 需要 root 权限运行: sudo python3 hello-tail.py")
    sys.exit(1)

# 解析参数
CHAIN_LEN = 3  # 默认完整链: entry -> handler1 -> handler2
LOAD_FALLBACK = True

for arg in sys.argv[1:]:
    if arg.startswith("--chain="):
        try:
            CHAIN_LEN = int(arg.split("=")[1])
            if CHAIN_LEN < 1 or CHAIN_LEN > 3:
                print("⚠️  chain 长度必须为 1-3，使用默认值 3")
                CHAIN_LEN = 3
        except ValueError:
            print("⚠️  参数格式错误，使用默认值")
    elif arg == "--no-fallback":
        LOAD_FALLBACK = False

print("=" * 60)
print("🧪 实验 2: eBPF 尾调用 (Tail Call)")
print("=" * 60)
print(f"   链长度: {CHAIN_LEN} 级")
print(f"   降级程序: {'✅' if LOAD_FALLBACK else '❌'}")
print()

# ============================================================
# 步骤 1: 加载 eBPF 程序
# ============================================================
print("[1/4] 加载 eBPF 程序...")

# 方法: 用 BCC 加载 C 源码
b = BPF(src_file="hello-tail.c")

# ============================================================
# 步骤 2: 获取各函数 fd
# ============================================================
print("[2/4] 获取函数文件描述符...")

entry_fn = b.load_func("hello_entry", BPF.KPROBE)
handler1_fn = b.load_func("hello_handler1", BPF.KPROBE)
handler2_fn = b.load_func("hello_handler2", BPF.KPROBE)
fallback_fn = b.load_func("hello_fallback", BPF.KPROBE)

print(f"   entry.fd    = {entry_fn.fd}")
print(f"   handler1.fd = {handler1_fn.fd}")
print(f"   handler2.fd = {handler2_fn.fd}")
print(f"   fallback.fd = {fallback_fn.fd}")

# ============================================================
# 步骤 3: 配置 prog_array (jmp_table)
# ============================================================
print("[3/4] 配置尾调用跳转表...")

jmp_table = b.get_table("jmp_table")

# 清空跳转表（BCC 初始化后默认全 0）
# 注意: BCC 的 prog_array 需要设置值

# 索引 0: 保留给 entry 自身（BCC 的 kprobe 附加用）
# 索引 1: handler1
jmp_table[ct.c_int(1)] = ct.c_int(handler1_fn.fd)
print("   jmp_table[1] = handler1 ✅")

# 索引 2: handler2（如果链长度 >= 2）
if CHAIN_LEN >= 2:
    jmp_table[ct.c_int(2)] = ct.c_int(handler2_fn.fd)
    print("   jmp_table[2] = handler2 ✅")
else:
    print("   jmp_table[2] = (空) — 将触发 fallback")

# 索引 3: fallback（可选）
if LOAD_FALLBACK:
    jmp_table[ct.c_int(3)] = ct.c_int(fallback_fn.fd)
    print("   jmp_table[3] = fallback ✅")

# ============================================================
# 步骤 4: 附加到 kprobe
# ============================================================
print("[4/4] 附加 kprobe...")

b.attach_kprobe(
    event=b.get_syscall_fnname("execve"),
    fn_name="hello_entry"
)
print("   已附加到 sys_execve\n")

print("=" * 60)
print("🚀 运行中! 在另一个终端执行命令观察 tail call")
print("   (如: ls, cat /etc/passwd, ps aux)")
print()
print("观察方法:")
print("   终端2: sudo cat /sys/kernel/debug/tracing/trace_pipe")
print("   终端3: sudo bpftool prog list | grep hello")
print("   终端3: sudo bpftool map dump name jmp_table")
print()
print("按 Ctrl+C 停止")
print("=" * 60)

try:
    b.trace_print()
except KeyboardInterrupt:
    print("\n\n🧹 清理中...")

# 清理时倒序卸载
print("   分离 kprobe...")
b.detach_kprobe(event=b.get_syscall_fnname("execve"), fn_name="hello_entry")

print("   程序已卸载，实验结束。")
print()
print("📖 实验回顾:")
print("   修改 --chain 参数重新运行，观察 fallback 行为:")
print(f"   sudo python3 hello-tail.py --chain=2")
print(f"   sudo python3 hello-tail.py --chain=1")